Add code to bootstrap KNOWN_PEERS from DNS SRV records.

graydon opened the pull request.

jedmccaleb commented on docs/stellar-core_example.cfg:
```diff
@@ -101,6 +101,11 @@ KNOWN_PEERS=[
 "core-testnet2.stellar.org", 
 "core-testnet3.stellar.org"]
 
+# BOOTSTRAP_DNS_SRV (single string) default is '_stellar-core._tcp.stellar.org'
```
I don't think this should default to anything. It basically defaults to something if we put this in the example config. If you are running standalone you don't want it to go fetch a bunch of peers from us.

graydon replied:
I guess my thinking here was to move increasingly towards "does something sensible with minimal configuration". If we had a default history source (stellar.org's public archive) as well, then you could literally run a nonvalidating stellar-core w/o any configuration.

nullstyle replied:
:+1: for literally running a non-validating stellar-core w/o configuration

jedmccaleb replied:
I get how that is kinda cool but I don't really think SDF should be the default for all that stuff. It is in the example config so it is very easy for people to use. It also leads to unexpected behavior when you do know what is going on. "I didn't put anything in your history store why is it still fetching history!?"

graydon commented:
Rebased and removed the default to SDF DNS.

If you want to see it working, try:
~~~~
$ ./src/stellar-core --dnssrv _xmpp-client._tcp.gmail.com
2015-09-29T19:16:10.631 <startup> [] [default] INFO  Querying DNS for SRV record of _xmpp-client._tcp.gmail.com
2015-09-29T19:16:10.656 <startup> [] [default] INFO  DNS SRV record: xmpp.l.google.com:5222 (prio 5, weight 0)
2015-09-29T19:16:10.656 <startup> [] [default] INFO  DNS SRV record: alt4.xmpp.l.google.com:5222 (prio 20, weight 0)
2015-09-29T19:16:10.656 <startup> [] [default] INFO  DNS SRV record: alt2.xmpp.l.google.com:5222 (prio 20, weight 0)
2015-09-29T19:16:10.656 <startup> [] [default] INFO  DNS SRV record: alt3.xmpp.l.google.com:5222 (prio 20, weight 0)
2015-09-29T19:16:10.656 <startup> [] [default] INFO  DNS SRV record: alt1.xmpp.l.google.com:5222 (prio 20, weight 0)
Selected peer: xmpp.l.google.com:5222
~~~~

jedmccaleb commented:
the comment saying it uses us as default is still there

vogel commented:
I think we should use an external library for handling DNS (https://github.com/wahern/dns looks good).
Current version only works on Linux. When compiled on Windows, all the code is no-op.

Also tests for pickSrvRecordByPriorityAndWeight are needed and consistent error handling. Error message looks differently when resolving command line parameter --dnssrv and config file parameter BOOTSTRAP_DNS_SRV).

@jedmccaleb: Let me know if you prefer external library for dns, or if I should implement it separately for Windows or if SRV queries on Windows should not be supported.

MonsieurNicolas commented:
we're taking a different approach now
